go-controller/pkg/ovn/libovsdbops/router.go

```diff
@@ -0,0 +1,312 @@
+package libovsdbops
+
+import (
+	"fmt"
+	"net"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	"github.com/ovn-org/libovsdb/model"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+func buildRouterNAT(
+	natType nbdb.NATType,
+	externalIP string,
+	logicalIP string,
+	logicalPort string,
+	externalMac string,
+	externalIDs *map[string]string,
+) *nbdb.NAT {
+	nat := &nbdb.NAT{
+		Type:       natType,
+		ExternalIP: externalIP,
+		LogicalIP:  logicalIP,
+		Options:    map[string]string{"stateless": "false"},
+	}
+
+	if logicalPort != "" {
+		nat.LogicalPort = &logicalPort
+	}
+
+	if externalMac != "" {
+		nat.ExternalMAC = &externalMac
+	}
+
+	if externalIDs != nil {
+		nat.ExternalIDs = *externalIDs
+	}
+
+	return nat
+}
+
+func BuildRouterSNAT(
+	externalIP *net.IP,
+	logicalIP *net.IPNet,
+	logicalPort string,
+	externalIDs *map[string]string,
+) *nbdb.NAT {
+	externalIPStr := ""
+	if externalIP != nil {
+		externalIPStr = externalIP.String()
+	}
+	// Strip out mask of logicalIP only if it is a host mask
+	logicalIPMask, _ := logicalIP.Mask.Size()
+	logicalIPStr := logicalIP.IP.String()
+	if logicalIPMask != 32 && logicalIPMask != 128 {
+		logicalIPStr = logicalIP.String()
+	}
+	return buildRouterNAT(nbdb.NATTypeSNAT, externalIPStr, logicalIPStr, logicalPort, "", externalIDs)
+}
+
+func BuildRouterDNATAndSNAT(
```

Comment:
This API looks a bit contrived: I think it would be a bit better if we had `BuildRouterNAT` which accepts a `nbdb.NATType` arg, it would reduce the amount of functions, leading to a flatter API.

Reply:
I was struggling with that a little bit. The problem here is that SNAT and DNAT types have a different handling of how it puts together the logical and external ips. My solution was to have the non-common pieces done as is then then invoke that part that is common to both of them ( i.e. buildRouterNAT ) . Would you like it better if I made buildRouterNAT a public function (as in BuildRouterNAT) ?

Reply:
Yeah, but I am not that familiar with why the logical and external IP would change depending on the type of the NAT object. So maybe if there are valid reasons for it, as you are aware, you can keep it like this for now.